infrastructure/templates/common/components/containerisedFunctionApp.bicep

```diff
@@ -0,0 +1,348 @@
+        }
+      })
+  )
+}
+
+module privateEndpointModule '../../public-api/components/privateEndpoint.bicep' = if (privateEndpoints.?functionApp != null) {
+  name: '${functionAppName}PrivateEndpointDeploy'
+  params: {
+    serviceId: functionApp.id
+    serviceName: functionApp.name
+    serviceType: 'sites'
+    subnetId: privateEndpoints.?functionApp ?? ''
+    location: location
+    tagValues: tagValues
+  }
+}
+
+module azureAuthentication '../../public-api/components/siteAzureAuthentication.bicep' = if (entraIdAuthentication != null) {
+  name: '${functionAppName}AzureAuthentication'
+  params: {
+    clientId: entraIdAuthentication!.appRegistrationClientId
+    siteName: functionAppName
+    allowedClientIds: entraIdAuthentication!.allowedClientIds
+    allowedPrincipalIds: entraIdAuthentication!.allowedPrincipalIds
+    requireAuthentication: entraIdAuthentication!.requireAuthentication
+  }
+}
+
+module healthAlert '../../public-api/components/alerts/staticMetricAlert.bicep' = if (alerts != null && alerts!.functionAppHealth) {
+  name: '${functionAppName}HealthAlertModule'
+  params: {
+    resourceName: functionAppName
+    resourceMetric: {
+      resourceType: 'Microsoft.Web/sites'
+      metric: 'HealthCheckStatus'
+    }
+    config: {
+      ...staticAverageLessThanHundred
+      nameSuffix: 'health'
+    }
+    alertsGroupName: alerts!.alertsGroupName
+    tagValues: tagValues
+  }
+  dependsOn: [
+    functionApp
+  ]
+}
+
+var unexpectedHttpStatusCodeMetrics = ['Http401', 'Http5xx']
+
+module unexpectedHttpStatusCodeAlerts '../../public-api/components/alerts/staticMetricAlert.bicep' = [
+  for httpStatusCode in unexpectedHttpStatusCodeMetrics: if (alerts != null && alerts!.httpErrors) {
+    name: '${functionAppName}${httpStatusCode}Module'
+    params: {
+      resourceName: functionAppName
+      resourceMetric: {
+        resourceType: 'Microsoft.Web/sites'
+        metric: httpStatusCode
+      }
+      config: {
+        ...staticMinGreaterThanZero
+        nameSuffix: toLower(httpStatusCode)
+      }
+      alertsGroupName: alerts!.alertsGroupName
+      tagValues: tagValues
+    }
+    dependsOn: [
+      functionApp
+    ]
+  }
+]
+
+var expectedHttpStatusCodeMetrics = ['Http403', 'Http4xx']
+
+module expectedHttpStatusCodeAlerts '../../public-api/components/alerts/dynamicMetricAlert.bicep' = [
+  for httpStatusCode in expectedHttpStatusCodeMetrics: if (alerts != null && alerts!.httpErrors) {
+    name: '${functionAppName}${httpStatusCode}Module'
+    params: {
+      resourceName: functionAppName
+      resourceMetric: {
+        resourceType: 'Microsoft.Web/sites'
+        metric: httpStatusCode
+      }
+      config: {
+        ...dynamicAverageGreaterThan
+        nameSuffix: toLower(httpStatusCode)
+        severity: 'Informational'
+      }
+      alertsGroupName: alerts!.alertsGroupName
+      tagValues: tagValues
+    }
+    dependsOn: [
+      functionApp
+    ]
+  }
+]
```

Comment:
I was thinking that `components/containerRegistryRoleAssignment.bicep` would also be necessary here with the principal id of the function app's user assigned or system assigned managed identity to make an 'AcrPull' role assignment for the function app on the container registry, giving the function app permission to pull from the container registry.

If you look at 'Role Assignments' tab under Access Control (IAM) within the 'eesacr' resource then you will see that the dedicated user assigned managed identity for the Public API's API app has this role assignment.

@duncan-at-hiveit do you agree with this because I noticed that in publicApiAppIdentity.bicep this code is commented out?

Reply:
So at the moment I believe that we aren't using identity-based ACR pulling, but instead are invoking the magical `DOCKER_*` environment variables instead along with the SPN details.

Mark's work on breaking out the ACR into a common Resource Group is where I think we're planning to attempt to make the switch to identity-based pulling.

Reply:
Ah yes, username/password authentication on the registry. @tomjonesdev I don't think you've got anything in to support that so maybe take a look at how `publicApiApp.bicep` sets params `dockerPullManagedIdentityClientId` and `dockerPullManagedIdentitySecretValue`, and how those params are used within `containerApp`.

The configuration could be completely different though because the Public API app is an Azure Container App in an Azure Container App environment, but you are writing an Azure Function that is not using the Azure Container App environment stack.